tests/contrib/celery/test_integration.py

```diff
@@ -694,6 +703,57 @@ def fn_task_parameters(user, force_logout=False):
         assert run_span.get_tag("component") == "celery"
         assert run_span.get_tag("span.kind") == "consumer"
 
+    def test_beat_scheduler_tracing(self):
+        @self.app.task
+        def fn_task():
+            return 42
+
+        target_task_run_count = 2
+        run_time_seconds = 1
+
+        self.app.conf.beat_schedule = {
+            "mytestschedule": {
+                "task": "tests.contrib.celery.test_integration.fn_task",
+                "schedule": run_time_seconds / target_task_run_count,
+            }
+        }
+
+        beat_service = celery.beat.EmbeddedService(self.app, thread=True)
+        beat_service.start()
+        sleep(run_time_seconds + 0.3)
+        beat_service.stop()
+
+        traces = self.pop_traces()
+        assert len(traces) >= 30  # tick() runs a large, unpredictable amount of times
+        assert traces[0][0].name == "celery.beat.tick"
+
+        # the following code verifies a trace structure in which every root span is either "celery.beat.tick" or
+        # "celery.run".
+        # some "celery.beat.tick" spans have no children, indicating that the beat scheduler checked for "due"
+        # tasks and found none.
+        # some "celery.beat.tick" spans have two children, "celery.beat.apply_entry" and "celery.apply".
+        # "celery.beat.apply_entry" is celery.beat's wrapper around apply_async, which triggers "celery.apply"
+        # and an asynchronous "celery.run" call.
+        # these "deep traces" indicate tick() calls that found a "due" task and triggered it
+
+        spans_counter = Counter()
+        deep_traces_count = 0
+        for trace in traces:
+            span_name = trace[0].name
+            spans_counter.update([span_name])
+            if span_name == "celery.beat.tick" and len(trace) > 1:
+                deep_traces_count += 1
+                spans_counter.update([trace[1].name, trace[2].name])
+                assert trace[1].name == "celery.beat.apply_entry"
+                assert trace[2].name == "celery.apply"
+        assert deep_traces_count >= target_task_run_count
+        assert target_task_run_count <= spans_counter["celery.run"] <= deep_traces_count
+        # beat_service.stop() can happen any time during the beat thread's execution.
+        # When by chance it happends between apply_entry() and run(), the number of spans
+        # for apply() and apply_entry() will be one greater than the number of spans for run()
```

Comment:
Am I correct in understanding that `celery.run()` will result in a `deep_traces_count += 1`? If by chance the `beat_service.stop()` happens between `apply_entry()` and `run()`, does that mean the service stops before the corresponding deep trace can be created?

Reply:
> If by chance the beat_service.stop() happens between apply_entry() and run(), does that mean the service stops before the corresponding deep trace can be created?

Almost. Because `run()` is called asynchronously by celery, it's not part of the "deep traces" this test checks for. Instead, it's a separate trace. This means that there's a moment between the deep trace ending and the run() trace starting during which the Beat thread can exit that will result in the run() trace being omitted.

I've changed the test logic a bit to try to make things more clear.